back/app/models/permission.rb

```diff
@@ -42,17 +42,29 @@ class Permission < ApplicationRecord
   belongs_to :permission_scope, polymorphic: true, optional: true
   has_many :groups_permissions, dependent: :destroy
   has_many :groups, through: :groups_permissions
-  has_many :permissions_custom_fields, -> { includes(:custom_field).order('custom_fields.ordering') }, inverse_of: :permission, dependent: :destroy
-  has_many :custom_fields, -> { order(:ordering) }, through: :permissions_custom_fields
+  has_many :permissions_fields, -> { includes(:custom_field).order('custom_fields.ordering') }, inverse_of: :permission, dependent: :destroy
+  has_many :custom_fields, -> { order(:ordering) }, through: :permissions_fields
 
   validates :action, presence: true, inclusion: { in: ->(permission) { available_actions(permission.permission_scope) } }
   validates :permitted_by, presence: true, inclusion: { in: PERMITTED_BIES }
-  validates :action, uniqueness: { scope: %i[permission_scope_id permission_scope_type] }
+  validates :action, uniqueness: { scope: %i[permission_scope_id permission_scope_type] }, if: -> { action != 'visiting' } # TODO: JS - add another validation for visiting
   validates :permission_scope_type, inclusion: { in: SCOPE_TYPES }
 
   before_validation :set_permitted_by_and_global_custom_fields, on: :create
   before_validation :update_global_custom_fields, on: :update
 
+  def permissions_fields
```

Comment:
Did you consider rendering the non-custom permission fields dynamically? I'm definitely not saying it's a better approach, but I think it would be good to compare the pros and cons. In this case, it looks like it could be useful to return different fields depending on the different conditions.

Reply:
Yes did that originally, but then you get into problems with includes / pagination etc if the fields don't actually exist and we need to include the custom_fields